src/components/DuplicateCourseModal.js

```diff
@@ -0,0 +1,162 @@
+      registrarId: courseListing.registrarId,
+      servicepointId: courseListing.servicepointId,
+      termId: term,
+    } }).json();
+
+    for (let i = 0; i < courseListing.instructorObjects.length; i++) {
+      await ky.post(`coursereserves/courselistings/${newCourseListing.id}/instructors`, { json: {
+        courseListingId: newCourseListing.id,
+        name: courseListing.instructorObjects[i].name,
+        userId: courseListing.instructorObjects[i].userId,
+      } }).json();
+    }
+
+    for (let i = 0; i < reserves.length; i++) {
+      await ky.post(`coursereserves/courselistings/${newCourseListing.id}/reserves`, { json: {
+        courseListingId: newCourseListing.id,
+        itemId: reserves[i].itemId,
+      } }).json();
+    }
+
+    const newCourse = await ky.post(`coursereserves/courselistings/${newCourseListing.id}/courses`, { json: {
+      courseListingId: newCourseListing.id,
+      courseNumber: course.courseNumber,
+      departmentId: course.departmentId,
+      description: course.description,
+      name: `${course.name} - ${intl.formatMessage({ id: 'ui-courses.button.duplicate' })}`,
+      sectionName: course.sectionName,
+    } }).json();
+
+    if (duplicateCrosslistedCourses) {
+      for (let i = 0; i < crossListed.length; i++) {
+        await ky.post(`coursereserves/courselistings/${newCourseListing.id}/courses`, { json: {
+          courseListingId: newCourseListing.id,
+          courseNumber: crossListed[i].courseNumber,
+          departmentId: crossListed[i].departmentId,
+          description: crossListed[i].description,
+          name: `${crossListed[i].name} - ${intl.formatMessage({ id: 'ui-courses.button.duplicate' })}`,
+          sectionName: crossListed[i].sectionName,
+        } }).json();
+      }
+    }
+
+    history.push(`/cr/courses/${newCourse.id}`);
+
+    onClose();
+  };
+
+  return (
+    <Modal
+      dismissible
+      id="duplicate-course-modal"
+      label={<FormattedMessage id="ui-courses.duplicateCourse" />}
+      onClose={onClose}
+      open
+    >
+      <Select
+        autoFocus
+        dataOptions={terms}
+        fullWidth
+        data-test-select-term-for-duplicate-course
+        label={<FormattedMessage id="ui-courses.field.term" />}
+        onChange={e => setTerm(e.target.value)}
+        required
+        value={term}
+      />
+      <Checkbox
+        checked={duplicateCrosslistedCourses}
+        data-test-checkbox-duplicate-crosslisted-courses
+        label={<FormattedMessage id="ui-courses.duplicateCourse.duplicateCrosslistedCourses" />}
+        onChange={e => setDuplicateCrosslistedCourses(e.target.checked)}
+      />
+      <Button
+        disabled={term === undefined}
+        fullWidth
+        data-test-duplicate-modal-clickable-duplicate
+        onClick={duplicateCourse}
+      >
+        <FormattedMessage id="ui-courses.button.createDuplicateCourses" />
+      </Button>
+    </Modal>
+  );
+};
+
+const CoursePropType = PropTypes.shape({
```

Comment:
I wonder if this is useful enough to be worth moving to its own file?

Reply:
I thought about it. I tend to default to the three-uses-before-generalization rule so decided against it - especially since it's a bit of a novel pattern - but I could see the argument. I think it's good for you to call it out so it sticks in our mind a bit the next time we find ourselves doing something similar so that we can think about whether it makes sense _then_. 

That said, I'm also generally not a fan of imported prop-type shapes because it lets you resolve newly-introduced prop-type errors with one swat rather than actually looking at each component that uses them. Let's say you have five different uses of this `CoursePropType`. Then we move `departmentId` to the courseListing on the backend. Prop type checks start failing so we fix the proptype and four of the places we use the department. But we missed one spot because This Is Programming and now we have a spot in the code that expects the old API and no longer warns about it.

Kind of reminds me of Defensive Driving...Defensive Programming 😆 

Of course, there's also a upside to explicitly having an `import CoursePropType` statement which actually makes it easier to _find_ usages of the Course shape that you should update when the API changes... 🤷

Reply:
This comment is a lovely demonstration of how programming is a much art as it is science. Over all, though, I am persuaded by your argument that importing prop-type shapes is generally a bad idea. Prop-type checking is close to worthless, but the process of being forced to _think_ about prop-types is valuable. We shouldn't short-circuit it.